crates/cargo-test-support/src/registry.rs

```diff
@@ -357,195 +419,323 @@ const DEFAULT_MODE: u32 = 0o644;
 
 /// Initializes the on-disk registry and sets up the config so that crates.io
 /// is replaced with the one on disk.
-pub fn init() {
-    let config = paths::home().join(".cargo/config");
-    if config.exists() {
-        return;
-    }
-    RegistryBuilder::new().build();
+pub fn init() -> TestRegistry {
+    RegistryBuilder::new().build()
 }
 
-/// Variant of `init` that initializes the "alternative" registry.
-pub fn alt_init() {
-    RegistryBuilder::new().alternative(true).build();
+/// Variant of `init` that initializes the "alternative" registry and crates.io
+/// replacement.
+pub fn alt_init() -> TestRegistry {
+    init();
+    RegistryBuilder::new().alternative().build()
 }
 
-pub struct RegistryServer {
-    done: Arc<AtomicBool>,
-    server: Option<thread::JoinHandle<()>>,
+pub struct HttpServerHandle {
+    join_handle: Option<thread::JoinHandle<()>>,
     addr: SocketAddr,
 }
 
-impl RegistryServer {
-    pub fn addr(&self) -> SocketAddr {
-        self.addr
+impl HttpServerHandle {
+    pub fn index_url(&self) -> Url {
+        Url::parse(&format!("sparse+http://{}/index/", self.addr.to_string())).unwrap()
+    }
+
+    pub fn api_url(&self) -> Url {
+        Url::parse(&format!("http://{}/", self.addr.to_string())).unwrap()
+    }
+
+    pub fn dl_url(&self) -> Url {
+        Url::parse(&format!("http://{}/dl", self.addr.to_string())).unwrap()
     }
 }
 
-impl Drop for RegistryServer {
+impl Drop for HttpServerHandle {
     fn drop(&mut self) {
-        self.done.store(true, Ordering::SeqCst);
-        // NOTE: we can't actually await the server since it's blocked in accept()
-        let _ = self.server.take();
+        if let Ok(mut stream) = TcpStream::connect(self.addr) {
+            // shutdown the server
+            stream.write_all(b"stop").unwrap();
+            stream.flush().unwrap();
+        }
+        let _ = self.join_handle.take().unwrap().join();
```

Comment:
Can these unwraps be removed?  I'm concerned that if something goes wrong, it risks a panic-in-panic situation which causes an abort and can be difficult to diagnose.  For example, if a test fails, the panic exiting the test will drop the handle and then if one of these fails, it would cause an abort.

Also, is there a particular need to join with the server thread?  I would think it should be fine to assume the thread will exit (and to possibly avoid a deadlock if the server thread is somehow stuck).  I think the original code I wrote included `.join()` just to ensure that everything was OK. But in the context of a drop handler, I'd be a little uncomfortable with it since it will get called even on a failed test.  We're mostly concerned with `cargo` executing correctly, and not too concerned with the server.

Reply:
The `join` was there to ensure the server was actually shutting down, but I can remove it.